web_services/foxx/brightid/db.js

```diff
@@ -319,168 +166,145 @@ function invite(inviter, invitee, groupId, data, timestamp) {
-    throw new errors.NotAdminError();
-  }
   group.admins.push(admin);
   groupsColl.update(group, { admins: group.admins });
 }
 
 function addUserToGroup(groupId, key, timestamp) {
-  const user = 'users/' + key;
-  const group = 'groups/' + groupId;
+  const _from = 'users/' + key;
+  const _to = 'groups/' + groupId;
 
-  const edge = usersInGroupsColl.firstExample({
-    _from: user,
-    _to: group
-  });
+  const edge = usersInGroupsColl.firstExample({ _from, _to });
   if (! edge) {
-    usersInGroupsColl.insert({
-      _from: user,
-      _to: group,
-      timestamp
-    });
+    usersInGroupsColl.insert({ _from, _to, timestamp });
   } else {
     usersInGroupsColl.update(edge, { timestamp });
   }
-
-}
-
-function addMembership(groupId, key, timestamp) {
-  if (! groupsColl.exists(groupId)) {
-    throw new errors.GroupNotFoundError(groupId);
-  }
-
+  // empty the group's vouchers after family group member changes
   const group = groupsColl.document(groupId);
-  if (group.isNew && ! group.founders.includes(key)) {
-    throw new errors.NewUserBeforeFoundersJoinError();
+  if (group.type == 'family') {
+    groupsColl.update(group, { vouchers: [] });
   }
+}
 
-  if (group.type == 'primary' && hasPrimaryGroup(key)) {
-    throw new errors.AlreadyHasPrimaryGroupError();
+function knownConnections(userId) {
+  return query`
+    FOR conn in ${connectionsColl}
+      FILTER conn._from == ${'users/' + userId}
+      FILTER conn.level IN ['already known', 'recovery']
+      FOR conn2 in ${connectionsColl}
+        FILTER conn2._from == conn._to
+        FILTER conn2._to == conn._from
+        FILTER conn2.level IN ['already known', 'recovery']
+      RETURN conn._to
+  `.toArray().map(id => id.replace('users/', ''));
+}
+
+function checkJoiningFamily(groupId, userId) {
+  const fg = familyGroup(userId);
+  if (fg && fg.head != userId) {
+    throw new errors.AlreadyIsFamilyMember();
+  }
+  const members = groupMembers(groupId);
+  const conns = knownConnections(userId);
+  // if some of group members are unknown
+  if (_.intersection(members, conns).length != members.length) {
+    throw new errors.IneligibleFamilyMember();
   }
+}
 
+function addMembership(groupId, key, timestamp) {
+  const group = getGroup(groupId);
   const invite = invitationsColl.firstExample({
     _from: 'users/' + key,
     _to: 'groups/' + groupId
   });
-  // invites will expire after 24 hours
-  if (!invite || timestamp - invite.timestamp >= 86400000) {
+  // invites will expire after 72 hours
+  if (!invite || timestamp - invite.timestamp >= 259200000) {
     throw new errors.NotInvitedError();
   }
   // remove invite after joining to not allow reusing that
   invitationsColl.remove(invite);
 
-  addUserToGroup(groupId, key, timestamp);
-
-  if (groupMembers(groupId).length == group.founders.length) {
-    groupsColl.update(group, { isNew: false });
+  if (group.type == 'family') {
+    checkJoiningFamily(groupId, key);
```

Comment:
This is already checked when the invite is created. No one can join a group without an invite, so it doesn't need to also be checked here.
```suggestion
```